### ISSUE TITLE
Bug: [require-await] missing error for function invocation

### ISSUE THREAD

**Issue by TkDodo** [typescript-eslint/typescript-eslint#7681]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.9.3&fileType=.tsx&code=MYewdgzgLgBAtgTwIIAcCWMC8MDeAoGGAMxBAC5cDCYBzAUygDFSKAKNMFAVygugCcONAJRYAfDAAK-EHDQQ6AOn50IIADYA3Ou049hAGioBfI8bx5QkWGrh0A8gCMAVnWCxs%2BQgHpvMKAAW8jAQASBc6gAmMHT8Mvwwjm4AhlwKMMkQCGDAMMFcYGBuqhDJ-AhU9EykAHIgANYUmdm5rKKYEoioaIokIIpVzCCsAOQjhlS%2B-kEQMXEg-LOgcW5Q6hWEg6T2jRlZOTBt4vDI6L2kAwxDeOYWQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRAR1ko9oEMA7v0r50URNGgB7aJHBgAviEVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```typescript
const myApi = {
  foo: {
    getFoo: (input: string) => Promise.resolve(input),
  },
}

const someObject = {
  // this should error because async is unnecessary
  getFooNok: async () => myApi.foo.getFoo(''),
  // this errors correctly
  getFooOk: async () => myApi.foo.getFoo
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/<rule-name>": ["error", ...<options>],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I'd expect `getFooNok` to show an error because the `async` keyword is unnecessary 

### Actual Result

`getFooNok` doesn't show an error

### Additional Info

I've encountered this issue while using trpc. The code is something like this:

```
t.router({
  foo: t.procedure
        .input(z.string())
        .query(async ({ input }) => myApi.getFoo('input')
})
```

here, too, the `async` keyword on the query should be shown as unnecessary, but it is not.

**Comment by bradzacher** [comment]
This is working as expected and is the entire point of our extension rule!

People like annotating functions that return promises as `async` and they didn't like that the base rule would force them to remove the `async`.

It is unnecessary but it's a nice bit of clear and explicit documentation.

This is mentioned in the rule docs
https://typescript-eslint.io/rules/require-await/

> This rule extends the base eslint/require-await rule. It uses type information to add support for async functions that return a Promise.

If you don't want this type-aware behaviour then you can turn off our extension rule and instead use the base rule.

**Comment by TkDodo** [comment]
🤦‍♂️ got it, thank you for clarifying

**Comment by JSteunou** [comment]
I had the same case and though about an issue with typescript-eslint too. This got me here, so I'm glad to find there was no issue, but maybe the documentation is unclear. I though all typescript-eslint rules that override eslint rules kept the same behavior.

**Comment by JoshuaKGoldberg** [comment]
> maybe the documentation is unclear

Agreed, the one sentence in https://typescript-eslint.io/rules/require-await isn't very explicit. Filed an issue: #7697. Thanks y'all!
